### PR TITLE
fix internal precision loss

### LINKED ISSUE #10488: In python3, rectangle does not appear when x axis is type datetime
I've been having trouble plotting rectangular glyphs with bokeh in Python 3. I can't find this issue documented anywhere, or any efforts to correct it.

I'm trying to plot rectangles whose x-axis extent is given as a timedelta, and whose positions are datetimes.

This is a minimum example that produces the error. I generate a set of times via numpy's datetime64, convert them to datetime objects, and plot the set of rectangles. ~No rectangles appear in the plot if the times are close to the current UTC, but if I shift them back 50 years they appear just fine.~ This was in error -- The rectangles appear, but their sizes vary. If I make them narrower by a factor of 10, they disappear altogether.

```
#!/bin/env python
from datetime import datetime, timedelta  
import numpy as np
from bokeh.plotting import figure, show
from bokeh.models import DatetimeTickFormatter
  
## The line below is the key -- If I subtract 50 years from the current date, the rectangles appear just fine.
## If i don't subtract them, the rectangles are not rendered.
t0 = datetime.utcnow() - timedelta(weeks=52 * 50)
 
times = t0.timestamp() * 1e3 + np.linspace(0, 5, 100, endpoint=False) * 3600 * 1e3
times = times.astype('datetime64[ms]')
  
plot = figure(
      output_backend='svg',
      plot_height=500,
      plot_width=1280,
      x_axis_label='UTC Time',
      y_axis_label="",
      y_axis_type='linear'
)
  
plot.xaxis.formatter = DatetimeTickFormatter(
      hours=['%H:%M'],
      minutes=['%H:%M'],
      seconds=['%H:%M']
)
  
dm = np.linspace(0, 100, 20)
width = (np.diff(times)[0] * 0.8).astype(timedelta)
times, dm = np.meshgrid(times, dm)
plot.rect(
      x=times.astype(datetime).flatten(),
      y=dm.flatten(), width=width,
      height=100 / 20, name='rect'
)
  
show(plot)
```

Note the indicated line right after the import statements -- it seems that bokeh is having trouble with recent dates in Python3, probably because their unix times in milliseconds are very large. If I shift back to 1970, as shown here, the error goes away.

I suspect that [this function](https://github.com/bokeh/bokeh/blob/b56543346bad8cdabe99960e63133619feaafbb7/bokeh/util/serialization.py#L162) may be part of the problem. The internal serialization in bokeh converts all datetimes to milliseconds since the unix epoch, which might be overflowing something.

### PR DESCRIPTION
**Pull request by bryevdv**
- [x] issues: fixes #10488
- [x] tests added / passed
- [ ] release document entry (if new feature or API change)

I could not find any other CenterRotatable that needed similar treatment, and I verified that hover works with just these changes. 

@mattpap any thoughts welcome:

* is it worth trying to test the array type to use `NumberArray` when possible instead of `Array` for everything in these places?
* Ideas for tests? I spent ~30 minutes trying to get a `Rect._map_data` unit test that worked but did not succeed

```
 #!/bin/env python
from datetime import datetime, timedelta
import numpy as np
from bokeh.plotting import figure, show
from bokeh.models import DatetimeTickFormatter

## The line below is the key -- If I subtract 50 years from the current date, the rectangles appear just fine.
## If i don't subtract them, the rectangles are not rendered.
t0 = datetime(2020, 9, 21, 23, 22, 25, 624793) # datetime.utcnow() #- timedelta(weeks=52 * 50)

times = t0.timestamp() * 1e3 + np.linspace(0, 5, 100, endpoint=False) * 3600 * 1e3
times = times.astype('datetime64[ms]')

plot = figure(
      output_backend='svg',
      plot_height=500,
      plot_width=1280,
      x_axis_label='UTC Time',
      y_axis_label="",
      y_axis_type='linear',
      tooltips="$index"
)

plot.xaxis.formatter = DatetimeTickFormatter(
      hours=['%H:%M'],
      minutes=['%H:%M'],
      seconds=['%H:%M']
)

dm = np.linspace(0, 100, 1)
width = 100000 # (np.diff(times)[0] * 0.8).astype(timedelta)
times, dm = np.meshgrid(times, dm)
plot.rect(
      x=list(times.astype(datetime).flatten()),
      y=dm.flatten(),
      width=width, height=100 / 20, name='rect',
      line_color="red", alpha=0.6
)

show(plot)
```
Screen Shot 2020-09-22 at 2 33 43 PM

### CODE REVIEW

File: bokehjs/src/lib/models/glyphs/glyph.ts [hunk 1]
```diff
@@ -142,17 +142,17 @@ export abstract class GlyphView extends View {
     const n = pts.length
     if (pts_location == 'center') {
       const halfspan = map(spans, (d) => d/2)
-      pt0 = new NumberArray(n)
+      pt0 = new Array(n)
```
Comment (mattpap) [inline bokehjs/src/lib/models/glyphs/glyph.ts:145]: For the sake of consistency I would use typed arrays exclusively, in this case `Float64Array`, though it won't have any measurable run time effect.

**Comment by mattpap** [comment]
Given the localized and private nature of those changes, I suppose this is a fine workaround, but in the long term we need to use the right word size for the data, even if it means that we will need to use synthetic coordinates and/or use more memory in webgl backend.

File: bokehjs/src/lib/models/glyphs/glyph.ts (hunk 1 above)
Comment (bryevdv) [reply bokehjs/src/lib/models/glyphs/glyph.ts:145]: Sure that seems reasonable

**Comment by bryevdv** [comment]
@mattpap this has nothing to do with webgl, as we discussed previously, we are doing in all the mapping in non-webgl now, which means that webgl only sees things that fit in float32 (e.g. screen coords). More pointedly, this problem surfaced on non-webgl backends. The issue is specifically wrt to timestamps, which do not fit sufficiently in 32 bit types. I am happy to consider any breaking changes for this for 3.0 (that is what #10379 is for hashing out)

**Comment by bryevdv** [comment]
@mattpap It think the BokehJS-CI tests just need update, the actual code from the issue in question looks correct:

Screen Shot 2020-09-22 at 4 49 49 PM

i.e. probably  a bunch of fractional pixel diffs.

**Comment by bryevdv** [comment]
Marked as ready. Happy to add a test but would need some guidance.

**Pull request merged**